cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -0,0 +1,273 @@
+    Thus, this class is should be instantiated using a `DeviceSpecification` proto via the
+    `from_proto()` class method.
+    """
+
+    def __init__(self, metadata: cirq.GridDeviceMetadata):
+        """Creates a GridDevice object.
+
+        This constructor typically should not be used directly. Use `from_proto()` instead.
+        """
+        self._metadata = metadata
+
+    @classmethod
+    def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GridDevice':
+        """Create a `GridDevice` from a DeviceSpecification proto.
+
+        Args:
+            proto: The `DeviceSpecification` proto describing a Google device.
+
+        Raises:
+            ValueError: If the given `DeviceSpecification` is invalid. It is invalid if:
+                * A `DeviceSpecification.valid_qubits` string is not in the form `<int>_<int>`, thus
+                  cannot be parsed as a `cirq.GridQubit`.
+                * `DeviceSpecification.valid_targets` refer to qubits which are not in
+                  `DeviceSpecification.valid_qubits`.
+                * A target set in `DeviceSpecification.valid_targets` has type `SYMMETRIC` or
+                  `ASYMMETRIC` but contains targets with repeated qubits, e.g. a qubit pair with a
+                  self loop.
+                * A target set in `DeviceSpecification.valid_targets` has type `SUBSET_PERMUTATION`
+                  but contains targets which do not have exactly one element. A `SUBSET_PERMUTATION`
+                  target set uses each target to represent a single qubit, and a gate can apply to
+                  any subset of qubits in the target set.
+
+        """
+
+        _validate_device_specification(proto)
+
+        # Create qubit set
+        all_qubits = [v2.grid_qubit_from_proto_id(q) for q in proto.valid_qubits]
+
+        # Create qubit pair set
+        #
+        # While the `GateSpecification` proto message contains qubit target references, they are
+        # ignored here because the following assumptions make them unnecessary currently:
+        # * All valid qubit pairs work for all two-qubit gates.
+        # * All valid qubits work for all single-qubit gates.
+        # * Measurement gate can always be applied to all subset of qubits.
+        #
+        qubit_pairs = [
+            (v2.grid_qubit_from_proto_id(target.ids[0]), v2.grid_qubit_from_proto_id(target.ids[1]))
+            for ts in proto.valid_targets
+            for target in ts.targets
+            if len(target.ids) == 2 and ts.target_ordering == v2.device_pb2.TargetSet.SYMMETRIC
+        ]
+
+        # TODO(#5050) implement gate durations
+        try:
+            metadata = cirq.GridDeviceMetadata(
+                qubit_pairs=qubit_pairs,
+                gateset=cirq.Gateset(),  # TODO(#5050) implement
+                all_qubits=all_qubits,
+            )
+        except ValueError as ve:  # coverage: ignore
```

Comment:
I added them because these lines are currently unreachable if everything works correctly, and only act as a last line of defense.

Reply:
Can we add a test which gives a bad input that triggers this condition?

Reply:
The same conditions that would trigger this exception would also cause `DeviceSpecification` validation to fail, so this is captured by the test `test_grid_device_invalid_device_specification`. If validation is buggy and this exception here is triggered, the test should get an exception but will fail because there's an error pattern mismatch.

Per @MichaelBroughton 's suggestion I added spec validation logic at the top of `from_proto` to make it explicit that errors that would've caused `GridDeviceMetadata` instantiation to fail are due to an invalid `DeviceSpecification`.

Will update the error match patterns to include "Invalid DeviceSpecification" to make sure that validation error messages are matched instead of the GridDeviceMetadata ones, in case error messages accidentally match up in the future between the two.

Reply:
Is it even worth having the ValueError catching if unreachable?

Reply:
I think so - it would protect against future bugs in validation logic. For example, when GridDeviceMetadata gets an additional constraint but validation isn't updated, it's useful to give users the hint that the error they are getting is due to the DeviceSpecification being invalid rather than a mistake they made.

Reply:
I also think we should remove this check since the code path is unreachable. 

We can add an assertion instead if we want to make sure we catch the inconsistencies between validation logic vs new constraints in GridDeviceMetadata but I'm not sure if putting everything under the umbrella of  device specification errors is the right thing to do in this case.